Index: linux-3.18.29/drivers/usb/serial/cp210x.c
===================================================================
--- linux-3.18.29.orig/drivers/usb/serial/cp210x.c
+++ linux-3.18.29/drivers/usb/serial/cp210x.c
@@ -52,6 +52,7 @@ static const struct usb_device_id id_tab
 	{ USB_DEVICE(0x0471, 0x066A) }, /* AKTAKOM ACE-1001 cable */
 	{ USB_DEVICE(0x0489, 0xE000) }, /* Pirelli Broadband S.p.A, DP-L10 SIP/GSM Mobile */
 	{ USB_DEVICE(0x0489, 0xE003) }, /* Pirelli Broadband S.p.A, DP-L10 SIP/GSM Mobile */
+	{ USB_DEVICE(0x054C, 0x06D9) }, /* USB 485MOD V1 by stack */
 	{ USB_DEVICE(0x0745, 0x1000) }, /* CipherLab USB CCD Barcode Scanner 1000 */
 	{ USB_DEVICE(0x0846, 0x1100) }, /* NetGear Managed Switch M4100 series, M5300 series, M7100 series */
 	{ USB_DEVICE(0x08e6, 0x5501) }, /* Gemalto Prox-PU/CU contactless smartcard reader */
